src/com/goide/inspections/GoPlaceholderChecker.java

```diff
@@ -0,0 +1,386 @@
+      this.format = format;
+      this.startPos = startPos;
+      this.argNum = argNum;
+    }
+
+    @NotNull
+    private Placeholder toPlaceholder() {
+      return new Placeholder(state, startPos, format, flags, argNums, verb);
+    }
+  }
+
+  @NotNull
+  private static FormatState parsePrintfVerb(@NotNull String format, int startPos, int argNum) {
+    FormatState state = new FormatState(format, startPos, argNum);
+
+    parseFlags(state);
+
+    if (!parseIndex(state)) return state;
+
+    // There may be a width
+    if (!parseNum(state)) return state;
+
+    if (!parsePrecision(state)) return state;
+
+    // Now a verb, possibly prefixed by an index (which we may already have)
+    if (!parseIndex(state)) return state;
+
+    if (state.nBytes == format.length()) {
+      state.state = Placeholder.State.MISSING_VERB_AT_END;
+      return state;
+    }
+
+    state.verb = PrintVerb.getByVerb(state.format.charAt(state.nBytes));
+    if (state.verb != null && state.verb != PrintVerb.Percent) state.argNums.add(state.argNum);
+    state.nBytes++;
+    state.format = state.format.substring(0, state.nBytes);
+    state.state = Placeholder.State.OK;
+
+    return state;
+  }
+
+  public static boolean hasPlaceholder(@NotNull String formatString) {
+    return StringUtil.containsChar(formatString, '%');
+  }
+
+  private static void parseFlags(@NotNull FormatState state) {
+    String knownFlags = "#0+- ";
+    StringBuilder flags = new StringBuilder(state.flags);
+    while (state.nBytes < state.format.length()) {
+      if (StringUtil.containsChar(knownFlags, state.format.charAt(state.nBytes))) {
+        flags.append(state.format.charAt(state.nBytes));
+      }
+      else {
+        state.flags = flags.toString();
+        return;
+      }
+      state.nBytes++;
+    }
+    state.flags = flags.toString();
+  }
+
+  private static void scanNum(@NotNull FormatState state) {
+    while (state.nBytes < state.format.length()) {
+      if (!StringUtil.isDecimalDigit(state.format.charAt(state.nBytes))) {
+        return;
+      }
+      state.nBytes++;
+    }
+  }
+
+  private static boolean parseIndex(@NotNull FormatState state) {
+    if (state.nBytes == state.format.length() || state.format.charAt(state.nBytes) != '[') return true;
+
+    state.indexed = true;
+    state.nBytes++;
+    int start = state.nBytes;
+    scanNum(state);
+    if (state.nBytes == state.format.length() || state.nBytes == start || state.format.charAt(state.nBytes) != ']') {
+      state.state = Placeholder.State.ARGUMENT_INDEX_NOT_NUMERIC;
+      return false;
+    }
+
+    int arg;
+    try {
+      arg = Integer.parseInt(state.format.substring(start, state.nBytes));
+    }
+    catch (NumberFormatException ignored) {
+      state.state = Placeholder.State.ARGUMENT_INDEX_NOT_NUMERIC;
+      return false;
+    }
+
+    state.nBytes++;
+    state.argNum = arg;
+
+    return true;
+  }
+
+  private static boolean parseNum(@NotNull FormatState state) {
```

Comment:
why is this always return `true`? and why other code supposes the function can return `false`?